File: standalone_app_Tear/pipeline.py
import logging
from typing import List, Tuple

import cv2
import numpy as np

from inference import InferenceResult, OnnxModel
from preprocess import PreSegmentation
from utils import Click, ModelImage

LOGGER = logging.getLogger(__name__)


class SurgicalPipeline:
    def __init__(self, onnx_dir: str):
        self.pre_model = PreSegmentation(f"{onnx_dir}/pre.onnx")
        self.first_model = OnnxModel(f"{onnx_dir}/first.onnx")
        self.iteration_model = OnnxModel(f"{onnx_dir}/iteration.onnx")

    def run_presegmentation(self, image_pil) -> Tuple[np.ndarray, np.ndarray]:
        pre_result = self.pre_model.infer(np.array(image_pil))
        return pre_result.area_mask, pre_result.faz_mask

    def run_iteration(
        self,
        image: ModelImage,
        clicks: List[Click],
        prev_logits: np.ndarray,
    ) -> InferenceResult:
        resized_hw = (image.resized_np.shape[0], image.resized_np.shape[1])
        mask_input_shape = None
        for shape in self.iteration_model.io.input_shapes.values():
            if len(shape) == 4 and shape[1] == 1:
                if shape[2] in (-1, None) or shape[3] in (-1, None):
                    mask_input_shape = None
                else:
                    mask_input_shape = (shape[2], shape[3])
                break
        if mask_input_shape is None:
            mask_input_shape = (resized_hw[0] // 4, resized_hw[1] // 4)

        mask_input = cv2.resize(
            prev_logits.astype(np.float32),
            (mask_input_shape[1], mask_input_shape[0]),
            interpolation=cv2.INTER_LINEAR,
        )[None, None, ...]

        return self.iteration_model.infer(
            image.resized_np,
            resized_hw=resized_hw,
            orig_hw=(image.original_np.shape[0], image.original_np.shape[1]),
            clicks=clicks,
            mask_input=mask_input,
        )

File: standalone_app_Tear_go_current_version/internal/pipeline/pipeline.go
package pipeline

import (
	"fmt"
	"image"
	"strings"

	"eyesam/standalone_app_Tear_go_current_version/internal/inference"
	"eyesam/standalone_app_Tear_go_current_version/internal/preprocess"
	"eyesam/standalone_app_Tear_go_current_version/internal/utils"
)

type Pipeline struct {
	PreModel       *preprocess.PreSegmenter
	FirstModel     *inference.OnnxModel
	IterationModel *inference.OnnxModel
}

func New(onnxDir string) *Pipeline {
	return &Pipeline{
		PreModel:       preprocess.NewPreSegmenter(fmt.Sprintf("%s/pre.onnx", onnxDir)),
		FirstModel:     inference.NewOnnxModel(fmt.Sprintf("%s/first.onnx", onnxDir)),
		IterationModel: inference.NewOnnxModel(fmt.Sprintf("%s/iteration.onnx", onnxDir)),
	}
}

func (p *Pipeline) RunPreSegmentation(img image.Image) (utils.Mask, utils.Mask, error) {
	modelImage := utils.PrepareImageForModel(img, img.Bounds().Dx(), img.Bounds().Dy())
	result, err := p.PreModel.Infer(modelImage)
	if err != nil {
		return utils.Mask{}, utils.Mask{}, err
	}
	return result.AreaMask, result.FAZMask, nil
}

func (p *Pipeline) RunFirst(image utils.ModelImage, clicks []utils.Click) (inference.Result, error) {
	return p.FirstModel.Infer(image, clicks, nil)
}

func (p *Pipeline) RunIteration(image utils.ModelImage, clicks []utils.Click, prevLogits utils.FloatMask) (inference.Result, error) {
	if prevLogits.Width == 0 || prevLogits.Height == 0 {
		return inference.Result{}, fmt.Errorf("prev logits missing")
	}
	if err := p.IterationModel.EnsureInitialized(); err != nil {
		return inference.Result{}, err
	}
	resizedHW := [2]int{image.Resized.Bounds().Dy(), image.Resized.Bounds().Dx()}
	maskInputShape := p.findMaskInputShape(resizedHW)
	maskInput := utils.ResizeFloatMaskLinear(prevLogits, maskInputShape[1], maskInputShape[0])
	return p.IterationModel.Infer(image, clicks, &maskInput)
}

func (p *Pipeline) findMaskInputShape(resizedHW [2]int) [2]int {
	for name, shape := range p.IterationModel.InputShapes {
		if strings.Contains(name, "mask") && len(shape) == 4 && shape[1] == 1 {
			h := int(shape[2])
			w := int(shape[3])
			if h > 0 && w > 0 {
				return [2]int{h, w}
			}
		}
	}
	return [2]int{resizedHW[0] / 4, resizedHW[1] / 4}
}
